csharp/src/lesson03/README.md

```diff
@@ -0,0 +1,327 @@
+using Microsoft.AspNetCore.Mvc;
+
+namespace OpenTracing.Tutorial.Lesson03.Exercise.Server.Controllers
+{
+    [Route("api/[controller]")]
+    public class FormatController : Controller
+    {
+        // GET: api/format
+        [HttpGet]
+        public string Get()
+        {
+            return "Hello!";
+        }
+
+        // GET: api/format/helloString
+        [HttpGet("{helloString}", Name = "GetFormat")]
+        public string Get(string helloString)
+        {
+            var formattedHelloString = $"Hello, {helloString}!";
+            return formattedHelloString;
+        }
+    }
+}
+```
+
+To have the server app on the port 8081, we have to adjust the `Properties/launchSettings.json` file. Make sure, 
+that all `applicationUrl` entries have the value `http://localhost:8081/`.
+
+Start the server app and access the endpoint directly:
+
+```powershell
+$ curl http://localhost:8081/api/format/Bryan
+```
+
+Executing the client still produces the same threes spans as in the previous lesson, meaning we
+have no traces for the server side:
+
+![Trace](trace-client.png)
+
+### Inter-Process Context Propagation
+
+Since the only change we made in the `Hello.cs` app was to replace two operations with HTTP calls,
+the tracing story remains the same - we get a trace with three spans, all from `hello-world` service.
+But now we have two more microservices participating in the transaction and we want to see them
+in the trace as well. In order to continue the trace over the process boundaries and HTTP calls,
+we need a way to propagate the span context over the wire. The OpenTracing API provides two functions
+in the Tracer interface to do that, `Inject(spanContext, format, carrier)` and `Extract(format, carrier)`.
+
+The `format` parameter refers to one of the three standard encodings the OpenTracing API defines:
+  * `TEXT_MAP` where span context is encoded as a collection of string key-value pairs,
+  * `BINARY` where span context is encoded as an opaque byte array,
+  * `HTTP_HEADERS`, which is similar to `TEXT_MAP` except that the keys must be safe to be used as HTTP headers.
+
+The `carrier` is an abstraction over the underlying RPC framework. For example, a carrier for `TEXT_MAP`
+format is an interface that allows the tracer to write key-value pairs via `Add(key, value)` method,
+while a carrier for Binary format is simply a `byte[]`.
+
+The tracing instrumentation uses `Inject` and `Extract` to pass the span context through the network calls.
+
+### Instrumenting the Client
+
+In the `FormatString` function we already create a child span. In order to pass its context over the HTTP
+request we need to call `_tracer.Inject` before building the HTTP request:
+
+```csharp
+var span = scope.Span
+    .SetTag("span.kind", "client")  // or: .SetTag(Tags.SpanKind, Tags.SpanKindClient);
+    .SetTag("http.method", "GET")   // or: .SetTag(Tags.HttpMethod, "GET");
+    .SetTag("http.url", url);       // or: .SetTag(Tags.HttpUrl, url);
```

Comment:
what's the reason to use strings instead of constants?

Reply:
To show that not only predefined values are usable but any string